builtin/logical/database/backend.go

```diff
@@ -228,6 +272,156 @@ func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage,
 	return db, nil
 }
 
+// initQueue preforms the necessary checks and initializations needed to preform
+// automatic credential rotation for roles associated with static accounts. This
+// method verifies if a queue is needed (primary server or local mount), and if
+// so initializes the queue and launches a go-routine to periodically invoke a
+// method to preform the rotations.
+//
+// initQueue is invoked by the Factory method in a go-routine. The Factory does
+// not wait for success or failure of it's tasks before continuing. This is to
+// avoid blocking the mount process while loading and evaluating existing roles,
+// etc.
+func (b *databaseBackend) initQueue(ctx context.Context, conf *logical.BackendConfig) {
+	// verify this mount is on the primary server, or is a local mount. If not, do
+	// not create a queue or launch a ticker. Both processing the WAL list and
+	// populating the queue are done sequentially and before launching a
+	// go-routine to run the periodic ticker.
+	replicationState := conf.System.ReplicationState()
+	if (conf.System.LocalMount() || !replicationState.HasState(consts.ReplicationPerformanceSecondary)) &&
+		!replicationState.HasState(consts.ReplicationDRSecondary) &&
+		!replicationState.HasState(consts.ReplicationPerformanceStandby) {
+		b.Logger().Info("initializing database rotation queue")
+
+		// Sleep a few seconds to allow Vault to mount and complete setup, so
+		// that we can write to storage
+		time.Sleep(3 * time.Second)
+
+		b.Lock()
+		if b.credRotationQueue == nil {
+			b.credRotationQueue = queue.New()
+		}
+		b.Unlock()
+
+		// create a context with a cancel method for processing any WAL entries and
+		// populating the queue
+		ctx, cancel := context.WithCancel(context.Background())
+		b.cancelQueue = cancel
```

Comment:
Should you move your Unlock after this assignment?

Reply:
Good catch thank you, done in c0907c0aa